grid2op/Chronics/GSFFWFWM.py

```diff
@@ -0,0 +1,148 @@
+        datelist = pd.date_range(self.start_datetime, periods=nbTimesteps, freq=freq)
+
+        datelist = np.unique(np.array([el.date() for el in datelist]))
+        datelist = datelist[:-1]
+
+        n_lines_maintenance = len(self.line_to_maintenance)
+
+        _24_h = timedelta(seconds=86400)
+        nb_rows = int(86400 / self.time_interval.total_seconds())
+        selected_rows_beg = int(self.maintenance_starting_hour * 3600 / self.time_interval.total_seconds())
+        selected_rows_end = int(self.maintenance_ending_hour * 3600 / self.time_interval.total_seconds())
+
+        res = np.zeros((nbTimesteps, len(self.name_line)))
+
+        # TODO this is INSANELY slow for now. find a way to make it faster
+        # HINT: vectorize everything into one single numpy array, everything can be vectorized there...
+        for nb_day_since_beg, this_day in enumerate(datelist):
+            dayOfWeek = this_day.weekday()
+            if dayOfWeek < 5:  # only maintenance starting on working days
+                month = this_day.month
+
+                maintenance_me = np.zeros((nb_rows, nb_line_maint))
+                # Careful: month start at 1 but inidces start at 0 in python
+                maintenance_daily_proba = self.daily_proba_per_month_maintenance[(month - 1)]
+                maxDailyMaintenance = self.max_daily_number_per_month_maintenance[(month - 1)]
+
+                # now for each line in self.line_to_maintenance, sample to know if we generate a maintenance
+                # for line in self.line_to_maintenance:
+                are_lines_in_maintenance = self.space_prng.choice([False, True],
+                                                                  p=[(1. - maintenance_daily_proba),
+                                                                     maintenance_daily_proba],
+                                                                  size=n_lines_maintenance)
+
+                n_Generated_Maintenance = np.sum(are_lines_in_maintenance)
+                # check if the number of maintenance is not above the max allowed. otherwise randomly pick up the right
+                # number
+                if (n_Generated_Maintenance > maxDailyMaintenance):
+                    # we pick up only maxDailyMaintenance elements
+                    not_chosen = self.space_prng.choice(n_Generated_Maintenance,
+                                                        replace=False,
+                                                        size=n_Generated_Maintenance - maxDailyMaintenance)
+                    are_lines_in_maintenance[np.where(are_lines_in_maintenance)[0][not_chosen]] = False
+                maintenance_me[selected_rows_beg:selected_rows_end, are_lines_in_maintenance] = 1.0
+
+                # handle last iteration
+                n_max = int( min(res.shape[0], (nb_day_since_beg + 1) * nb_rows) / (nb_day_since_beg + 1))
```

Comment:
My brain hurts, but i've seen the TODO at the top of the loop so it's fine for now

Reply:
You have not had to fix the original implementation :see_no_evil: 
Anyway best i can do in a reasonable amount of time :-/

But yes this has to be vectorized and optimized and refactor